week3/homework/src/index.js

```diff
@@ -1,3 +1,96 @@
 'use strict';
+const express = require('express');
+const uuid = require('uuid/v4');
 
-// TODO: Write the homework code in this file
+const { readFile: _readFile, writeFile: _writeFile } = require('fs');
+const { promisify } = require('util');
+
+const readFile = promisify(_readFile);
+const writeFile = promisify(_writeFile);
+
+const app = express();
+//build-in json middleware to automatically parse json
+app.use(express.json());
+const TODO_FILE = 'todos.json';
+
+function readTodos() {
+  return readFile(TODO_FILE, `utf8`).then(
+    JSON.parse,
+    () => ({})
+  );
+}
+
+function writeTodos(data) {
+  return writeFile(TODO_FILE, JSON.stringify(data));
+}
+
+//Create a todo
+app.post('/todos', async (req, res) => {
+  const newTodo = req.body;
+  const id = uuid();
+  newTodo.done = false;
+  const todos = await readTodos();
+  todos[id] = newTodo;
+  await writeTodos(todos);
+  res.json(todos);
+});
+
+//Read all todos
+app.get('/todos', async (req, res) => {
+  const todos = await readTodos();
+  res.json(todos);
+});
+
+//Read a todo by ID
+app.get('/todos/:id', async (req, res) => {
+  const id = req.params.id;
+  const todos = await readTodos();
+  res.json(todos[id]);
+});
+
+//Delete all todos
+app.delete('/todos', async (req, res) => {
+  await writeTodos({});
+  res.json({});
+});
+
+//Delete a todo by ID
+app.delete('/todos/:id', async (req, res) => {
+  const id = req.params.id;
+  const todos = await readTodos();
+  delete todos[id];
+  await writeTodos(todos);
+  res.json(todos);
+});
+
+//update a todo
+app.put('/todos/:id', async (req, res) => {
+  const id = req.params.id;
+  const newTodo = req.body;
+  const todos = await readTodos();
+  todos[id].description = newTodo.description;
```

Comment:
Great idea using IDs as keys. If you're curious about a more efficient way to do that, have a look at [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map).

Also you're not validating the input anywhere. What if the ID doesn't exist? In case of updating, you're going to create a new record even though you should have returned an error. Same thing for other routes that have an ID. Have a look at [app.param](https://expressjs.com/en/4x/api.html#app.param) for a way of validating individual params only once.

Reply:
Thank you I'll work more on it 👍